### ISSUE TITLE
scope resolving with method receiver scopes

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#21560]
This issue brings up a language design problem with what is visible from method receiver scopes. It discusses the problem through a series of examples that give related background.

### Example 1: Finding a Field

Within a method, an identifier might refer to a field. That field could be in a different module that declares the type. As a result, the method receiver scope (or scopes, for classes using inheritance), are consulted when scope resolving a method:

``` chapel
// Example 1
module M {
  record R {
    var x: int;
  }
}
module N {
  import M.R;
  proc R.foo() {
    x; // refers to the field R.x
  }
  proc main() { var r = new R(); r.foo(); }
}
```

That makes sense, at least as long as the field is not `private` (and we don't have `private` fields yet).

### Example 2: Finding a Secondary Method

Meanwhile, a method call within a method might refer to a secondary method defined in the module defining the type:

``` chapel
// Example 2
module M {
  record R { }
  proc R.bar() { }
}
module N {
  import M.R;
  proc R.foo() {
    bar(); // refers to M.R.bar()
  }
  proc main() { var r = new R(); r.foo(); }
}
```

That also makes sense. Otherwise, only primary methods would be available in such a circumstance, and it should not matter if a method is declared as primary or secondary.

OK, so far so good. Now, we also have the situation that a variable or parenless method does not participate in disambiguation. One way to view that is that the order in which the various scopes are considered matters, and the first scope that has a match will return that match.

### Example 3: Finding a Module-Scope Variable?

That brings us to this example:

``` chapel
// Example 3
module M {
  record R { }
  var moduleVar: int;
}
module N {
  import M.R;
  proc R.foo() {
    moduleVar; // refers to M.moduleVar (but, unresolved by dyno's scope resolver)
  }
  proc main() { var r = new R(); r.foo(); }
}
```

Is it reasonable that `moduleVar`, which is neither a field nor a method, is visible from `N.R.foo()`?

### Example 4: Finding a Module-Scope Function

Oddly enough, a similar case with a function does not resolve:

``` chapel
// Example 4
module M {
  record R { }
  proc baz() { }
}
module N {
  proc R.foo() {
    baz(); // error: unresolved call 'baz()', no functions named 'baz' found in scope
  }
}
```

This seems inconsistent. (I think it's this way because we have 2 different implementations of visibility; one for variables and another for functions; and the functions one is only considering methods as something that can be added by the definition-point rule.)

### Example 5: Inconsistency between Functions and Variables

The above inconsistency can also be viewed in terms of which of two variables/functions is chosen:

``` chapel
// Example 5
module M {
  record R { }
  var moduleVar: int = 1;
  proc baz() { writeln("M.baz"); }
}
module N {
  import M.R;
  var moduleVar: int = 2;
  proc baz() { writeln("N.baz"); }
  proc R.foo() {
    writeln(moduleVar); // refers to M.moduleVar
    baz();              // refers to N.baz
  }
  proc main() { var r = new R(); r.foo(); }
}
```

(Note that if `baz` were a method, such an example would result in an ambiguity error, since methods don't use a visibility rule in disambiguation).

### Example 6: Hijacking

The current behavior with a module-level variable could, arguably, be a source of "hijacking" behavior.

Suppose we start out with this library and application code (developed by different groups):

``` chapel
module LibraryV1 {
  record R { }
}
```

``` chapel
module Application {
  var x: int;
  proc R.helper() {
    writeln(x);
  } 
}
```

Then, suppose that Library is updated to have a module-level variable `x`:

``` chapel
module LibraryV2 {
  record R { }
  var x: int;
}
```

Now suppose that the Application us updated to use LibraryV2. It will no longer function as it did. Similarly to Example 5, the `x` within `Application.R.helper` will now refer to `LibraryV2.x` instead of `Application.x`.

### Proposal

I think it would be easy enough to include only fields and methods in the symbols that are visible from method receiver scopes. That would address this issue.

### History

This came up due to a `--dyno` test failure with the test:

    test/classes/deitz/class/scope_resolve_classes.chpl

This test has different behavior with `--dyno` and I am not so sure that the production compiler has the right behavior for this test. This test was added as a future (asking "what should this do?" in b072b498d97abac3824cfb453e6874b2a7851b35 but the .future file was removed in 59f791076b8438e9c63924e81daf33450ce178a4.

**Comment by cassella** [comment]
Example 3:

> Is it reasonable that moduleVar, which is neither a field nor a method, is visible from N.R.foo()?

I'd say no.  I'd be surprised anything in `M` would be visible in that `foo()` that wasn't made visible by a `use` or `import`.

Besides variables, types defined in `M` are also visible in `foo()`:

```chapel
// Example 3
module M {
  record S { };
  var moduleS: S;
  record R { }
  var moduleVar: int;
}
module N {
  import M.R;
  proc R.foo() {
    writeln(moduleVar); // refers to M.moduleVar (but, unresolved by dyno's scope resolver)
    writeln(moduleS);
    writeln(moduleS.type:string);

    var myS: S;
    writeln(myS.type:string);
    
  }
  proc main() { var r = new R(); r.foo(); }
}
```

```
0
()
S
S
```

**Comment by bradcray** [comment]
I agree with Michael's proposal and Paul's reaction.

**Comment by mppf** [comment]
> I think it would be easy enough to include only fields and methods in the symbols that are visible from method receiver scopes. That would address this issue.

If we did that, what about a nested class case?

``` chapel
module M {
  class outerClass {
    class aClass {
      var a: int;
    }
    proc foo() {
      aClass; // does this refer to nested aClass our module-scope one?
    }
  }

  class aClass {
    var b: int;
  }
}
```

Here, the nested class `class aClass` is not a method or a field. Should it be considered along with methods and fields for this purpose?

**Comment by lydia-duncan** [comment]
I think so, personally.  But it is in the scope of the class, so I would just phrase the rule as "we don't look outside the class/record scope for symbols unless we were in that scope to start" rather than making a list of symbol types that we look for

**Comment by mppf** [comment]
@lydia-duncan - are you saying that it should depend on whether `foo` is a primary or secondary method? Today, it compiles either way & refers to `outerClass.aClass`. I don't think we should make it depend on whether the method is written primary or secondary.

<details>

Variation of the example that uses a secondary method:

``` chapel
module M {
  class outerClass {
    class aClass {
      var a: int;
    }
  }
  proc outerClass.foo() {
    aClass; // does this refer to nested aClass our module-scope one?
  }

  class aClass {
    var b: int;
  }
}
```
</details>

**Comment by lydia-duncan** [comment]
No, I was trying to distinguish between primary/secondary and tertiary.  I would expect both primary and secondary to look in the module where the type was defined and for tertiary to stop at the class/record boundary

**Comment by mppf** [comment]
I'm still not understanding. It's the case today that, if I write a tertiary method, I can call other tertiary methods within it:

``` chapel
module DefinesType {
  public use ExtendsType;
  record R { }
}

module ExtendsType {
  public import DefinesType.R;

  proc R.foo() { writeln("ExtendsType.foo()"); }
}

module Application {
  import DefinesType.R;

  proc R.bar() { writeln("Application.bar()"); foo(); }

  proc main() {
    var x = new R();
    x.bar();
  }
}
```

> stop at the class/record boundary

The method search for `foo` does not stop after considering the direct contents of 
`record R`. (Doing so would cause it to only consider primary methods, at least with the way I understand those words).

I suppose it is debateable whether my `R.foo` is a secondary or tertiary method (it is arguably secondary, because it is visible to the module defining `R`). FWIW, I consider it a tertiary method that is visible in the module defining `R`.

Anyway, I think methods are special in this way; for fields / nested classes / etc; we only consider things declared directly within the class/record as visible through the method receiver.

**Comment by lydia-duncan** [comment]
Sorry, I was not meaning to imply that secondary methods would not be considered.

**Comment by lydia-duncan** [comment]
I would also consider `R.foo` a tertiary method (and I thought the definition was not about visible but solely definition location - e.g. tertiary methods are defined outside of the module where the type is defined, regardless of whether that module is then used in the module where the type is defined).

**Comment by bradcray** [comment]
> Here, the nested class class aClass is not a method or a field. Should it be considered along with methods and fields for this purpose?

I think it should.  If outerClass.foo() were a secondary or tertiary method, I think I'd want the same behavior, but would appreciate a warning about the ambiguity (and wouldn't mind one for the primary method case either).
